### DISCUSSION TITLE
Bridgeman Thermodynamic Equations

### DISCUSSION THREAD

**Discussion by rob5ives** [Cantera/enhancements#101] [Ideas]
Is anyone here familiar with the Bridgeman Thermodynamic Tables? I learned about them when making a Chapman-Jouguet detonation solver for use with Cantera. My solver is based on the methods outlined in the CEA theory document and that is where I first found out about the Bridgeman equations. Check out the wiki: https://en.wikipedia.org/wiki/Bridgman%27s_thermodynamic_equations

I was originally using the Shock and Detonation Toolbox from CalTech. The problem is that their solver is not very robust. It is possible to make a more robust solver but it requires an equilibrium speed of sound calculation. However, since it is defined as a derivative, a numerical finite difference approximation is normally used. This is not very robust. 

When looking for alternatives I decided to check out how CEA does detonation calculations since I've never had issues with it other than I cannot easily script it like Cantera. It turns out they use the equilibrium speed of sound but they don't have to use a finite difference approximation.

Basically, if you've solved for the equilibrium state, you already know some basic thermodynamic derivatives. Using the Bridgeman equations, you can then construct other derivatives. No approximations necessary other than the error from determining equilibrium.

Right now I have a hacked together solution entirely in Python. I basically had to recreate the minimization function that would be present in the C++ equilibrium solver. Do you think it would be worthwhile to include this in the C++ code and Solution object? It would add a powerful way of determining thermodynamic derivatives at equilibrium. This is useful for all sorts of problems. Especially in compressible fluid flow which is what I generally do.

Let me know what you all think.

**Comment by bryanwweber** [comment]
This is awesome! Can you share your Python script? This is actually something I happened to be looking at earlier this year for my class, but I didn't make much progress, so I'm glad to see someone actually doing it right!

**Reply by rob5ives** [reply]
Attaching the scripts at text files. If you look, all I did was implement the equations outlined in the CEA theory document. That also means I implemented the minimization equation outlined in CEA. I'm not familiar with what Cantera uses but I assume it is something similar. 

My hope is to have the Cantera Solution object do something similar to CEA and calculate three equilibrium derivatives analytically. There would probably have to be a check to ensure that the Solution object is at thermodynamic equilibrium. Someone could then use the Bridgeman tables to generate the derivative they desire. For example, the equilibrium speed of sound.
`
[detonation.txt](https://github.com/Cantera/enhancements/files/6565453/detonation.txt)
[thermo.txt](https://github.com/Cantera/enhancements/files/6565454/thermo.txt)

`
